ecs-cli/modules/clients/aws/route53/servicediscovery_client.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2015-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package route53 contains functions for working with the route53 APIs
+// that back ECS Service Discovery
+package route53
+
+import (
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+)
+
+// FindPrivateNamespace returns the ID(s) of the private namespace with the given name and vpc
+func FindPrivateNamespace(name, vpc string, config *config.CommandConfig) (*string, error) {
+	region := aws.StringValue(config.Session.Config.Region)
+	var nameMatch []*string
+	r53Client := newRoute53Client(config)
+	sdClient := newSDClient(config)
```

Comment:
@SoManyHs I've refactored the code so that the clients are explicitly created in the public/exported functions. I think this should satisfy your concerns.

I've kept the two public functions as they are:

```
FindPrivateNamespace(name, vpc string, config *config.CommandConfig) (*string, error)
FindPublicNamespace(name string, config *config.CommandConfig) (*string, error)
```

The two public functions have signatures which just take `CommandConfig`, instead of them taking a client as an argument or being methods on a client struct. Let me explain why I chose this design. I think that the consumers of this package shouldn't have to care or work with the sdk/client at all. All a consumer of this package wants to do is `FindPrivateNamespace`- why should it have to care that doing so requires both a `Route53Client` and a `ServiceDiscoveryClient`? The public functions take a `CommandConfig` as an argument, because this is what is needed to create clients, and the internal implementation of the function is what actually deals with the client.

Reply:
I agree that the a *customer* of the CLI should not care about implementation details of client instantiation etc; I disagree that the consumer of the package shouldn't care about anything but the public API because it's an internal package, and so the ones using the package are the maintainers. This all has already been addressed in aithal's [comment](https://github.com/aws/amazon-ecs-cli/pull/565/commits/b0437ab9324e7039404213d6bb095f770c8d79b2#r208594610), but just thought I'd add my two cents.